tests/test_transform_sequence_association.py

```diff
@@ -0,0 +1,203 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import pytest
+
+from conftest import available_frontends
+from loki.transform import transform_sequence_association
+from loki.module import Module
+from loki.ir import CallStatement
+from loki.visitors import FindNodes
+from loki.expression import Sum, IntLiteral, Scalar, Product
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_transform_scalar_notation(frontend):
+    fcode = """
+module mod_a
+    implicit none
+
+    type type_b
+        integer :: c
+        integer :: d
+    end type type_b
+
+    type type_a
+        type(type_b) :: b
+    end type type_a
+
+contains
+
+    subroutine main()
+
+        type(type_a) :: a
+        integer :: k, m, n
+
+        real    :: array(10,10)
+
+        call sub_a(array(1, 1), k)
+        call sub_a(array(2, 2), k)
+        call sub_a(array(m, m), k)
+        call sub_a(array(m-1, m-1), k)
+        call sub_a(array(a%b%c, a%b%c), k)
+
+        call sub_b(array(1, 1))
+        call sub_b(array(2, 2))
+        call sub_b(array(m, 2))
+        call sub_b(array(m-1, m), k)
+        call sub_b(array(a%b%c, 2))
+
+        call sub_c(array(1, 1), k)
+        call sub_c(array(2, 2), k)
+        call sub_c(array(m, 1), k)
+        call sub_c(array(m-1, m), k)
+        call sub_c(array(a%b%c, 1), k)
+
+        call sub_d(array(1, 1), 1, n)
+        call sub_d(array(2, 2), 1, n)
+        call sub_d(array(m, 1), k, n)
+        call sub_d(array(m-1, 1), k, n-1)
+        call sub_d(array(a%b%c, 1), 1, n)
+
+        call sub_e(array(1, 1), a%b)
+        call sub_e(array(2, 2), a%b)
+        call sub_e(array(m, 1), a%b)
+        call sub_e(array(m-1, 1), a%b)
+        call sub_e(array(a%b%c, 1), a%b)
+
+        call sub_x(array(1, 1), 1)
+        call sub_x(array(2, 2), 2)
+        call sub_x(array(m, 1), k)
+        call sub_x(array(m-1, 1), k-1)
+        call sub_x(array(a%b%c, 1), a%b%d)
```

Comment:
I was looking into it, but discovered that the code crashes if the dimension inside the routine is given by a Sum or a Product, for example `dimension(k+1)` and `dimension(-k)`, respectively.
The solution is to modify the `process_symbol` routine to recursively process such cases, but I'm wondering if I've been  really overthinking the problem. Since sequence association is always contiguous, I think we could just plug in the relevant array dimensions from the caller routine. At least the two calls in this example case produce the same output:
```
program crashdummies

    use blarg, only: blurg

    implicit none

    real, dimension(10,5) :: A
    integer :: j

    A = 0.

    call blurg(A(5:10,1))

    do j = 1,3
        print *, A(:,j)
    enddo
    print *

    A = 0.

    call blurg(A(5:19,1))

    do j = 1,3
        print *, A(:,j)
    enddo

end program crashdummies

module blarg

contains

    subroutine blurg(x)

        implicit none

        real, dimension(15) :: x

        integer :: i

        do i = 1, 15
            x(i) = 1.
        enddo

    end subroutine blurg

end module blarg
```

Would the inliner have any problem with this?

Reply:
Fantastic, simplicity is always better and this looks really neat now! Pylint has flagged a few redundant symbols and imports now.

I don't think the inliner should have a problem with this, at least there is a test that includes ranges: https://github.com/ecmwf-ifs/loki/blob/a2f2428e3f3ee3215f9dee90ae2dc438ba0cb1e3/tests/test_transform_inline.py#L374